test/cpp/api/serialize.cpp

```diff
@@ -33,6 +35,175 @@ torch::Tensor save_and_load(torch::Tensor input) {
+  torch::save(model1, model_tempfile.name);
+  torch::load(model2, model_tempfile.name);
+  torch::load(model3, model_tempfile.name);
+
+  auto param1 = model1->named_parameters();
+  auto param2 = model2->named_parameters();
+  auto param3 = model3->named_parameters();
+  for (const auto& p : param1) {
+    if(!p->allclose(param2[p.key()]) && param2[p.key()].allclose(param3[p.key()])) {
+      return false;
+    }
+  }
+  // Make some optimizers
+  auto optim1 = OptimizerClass(
+      {torch::optim::OptimizerParamGroup(model1->parameters())}, options);
+  auto optim2 = OptimizerClass(
+      model2->parameters(), options);
+  auto optim2_2 = OptimizerClass(
+      model2->parameters(), options);
+  auto optim3 = OptimizerClass(
+      model3->parameters(), options);
+  auto optim3_2 = OptimizerClass(
+      model3->parameters(), options);
+
+  auto x = torch::ones({10, 5});
+
+  auto step = [&x](torch::optim::Optimizer& optimizer, Linear model) {
+    optimizer.zero_grad();
+    auto y = model->forward(x).sum();
+    y.backward();
+    optimizer.step();
+  };
+
+  // Do 2 steps of model1
+  step(optim1, model1);
+  step(optim1, model1);
+
+  // Do 2 steps of model 2 without saving the optimizer
+  step(optim2, model2);
+  step(optim2_2, model2);
+
+  // Do 2 steps of model 3 while saving the optimizer
+  step(optim3, model3);
+
+  auto optim_tempfile = c10::make_tempfile();
+  torch::save(optim3, optim_tempfile.name);
+  torch::load(optim3_2, optim_tempfile.name);
+  step(optim3_2, model3);
+
+  const auto& optim3_2_param_groups = optim3_2.param_groups();
+  const auto& optim1_param_groups = optim1.param_groups();
+  const auto& optim3_2_state = optim3_2.state();
+  const auto& optim1_state = optim1.state();
+
+  //optim3_2 and optim1 should have param_groups and state of same size
+  if(!((optim3_2_param_groups.size() == optim1_param_groups.size()) &&
+     (optim3_2_state.size() == optim1_state.size()))) {
+    return false;
+  }
+
+  // checking correctness of serialization logic for optimizer.param_groups_ and optimizer.state_
+  for(int i=0; i<optim3_2_param_groups.size(); i++) {
+    if(!is_optimizer_param_groups_equal<DerivedOptimizerOptions>(
+      optim3_2_param_groups[i], optim1_param_groups[i])) {
+      return false;
+    }
+    if(!is_optimizer_state_equal<DerivedOptimizerParamState>(
+      optim3_2_param_groups[i].params(), optim3_2_state,
+      optim1_param_groups[i].params(), optim1_state)) {
+      return false;
+    }
+  }
+  param1 = model1->named_parameters();
+  param2 = model2->named_parameters();
+  param3 = model3->named_parameters();
+  for (const auto& p : param1) {
+    const auto& name = p.key();
+    // Model 1 and 3 should be the same
+    if(!((param1[name].norm().item<float>() == param3[name].norm().item<float>()) &&
+          (param1[name].norm().item<float>() != param2[name].norm().item<float>()))) {
+      return false;
+    }
+  }
+
+  // bc compatibility check
```

Comment:
We might want to put the BC check into another test function that's specific to `Adagrad`. The reason is that `sum_buffers` and `step_buffers` are specific to the `Adagrad` optimizer only, and we won't be able to reuse this part of the test for other optimizers.

Reply:
yeah I agree will do